File: app/src/main/java/com/bondex/ysl/battledore/base/BaseActivity.java
package com.bondex.ysl.battledore.base;

import android.arch.lifecycle.*;
import android.databinding.DataBindingUtil;
import android.databinding.ViewDataBinding;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.FragmentActivity;
import android.support.v7.app.AppCompatActivity;
import android.text.TextUtils;
import android.util.Log;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;
import com.bondex.ysl.battledore.main.MainViewModle;
import com.bondex.ysl.liblibrary.ui.IconText;
import com.bondex.ysl.liblibrary.utils.NoDoubleClickListener;
import com.bondex.ysl.liblibrary.utils.StatusBarUtil;

import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;

import com.bondex.ysl.battledore.R;


/**
 * date: 2019/5/27
 * Author: ysl
 * description:
 */
public abstract class BaseActivity<M extends BaseViewModle, B extends ViewDataBinding> extends AppCompatActivity {

    protected M viewModel;
    protected B binding;

    private ImageView ivBack;
    private TextView tvTitle;
    private IconText iconRight;

    protected MyclickListener listener = new MyclickListener();


    private Observer<Integer> msgObserver = new Observer<Integer>() {
        @Override
        public void onChanged(@Nullable Integer msg) {
            handleMessage(msg);
        }
    };
    private Observer<Boolean> refreshObserver = new Observer<Boolean>() {
        @Override
        public void onChanged(@Nullable Boolean isShow) {

            if (isShow) startLoading();
            else stopLoading();
        }
    };


    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        initData();
        binding = DataBindingUtil.setContentView(this, getReourceId());

        ivBack = binding.getRoot().findViewById(R.id.base_back);
        tvTitle = binding.getRoot().findViewById(R.id.base_title);
        iconRight = binding.getRoot().findViewById(R.id.base_right);

        if (ivBack != null) ivBack.setVisibility(View.INVISIBLE);
        if (tvTitle != null) tvTitle.setVisibility(View.INVISIBLE);
        if (iconRight != null) iconRight.setVisibility(View.INVISIBLE);

        viewModel = initViewModle();

        if (viewModel == null) {
            Class modelClass;
            Type type = getClass().getGenericSuperclass();
            if (type instanceof ParameterizedType) {
                modelClass = (Class) ((ParameterizedType) type).getActualTypeArguments()[0];
            } else {
                //如果没有指定泛型参数，则默认使用BaseViewModel
                modelClass = BaseViewModle.class;
            }
            viewModel = (M) createViewModel(this, modelClass);

            viewModel.setCont(getApplicationContext());
        }


        getLifecycle().addObserver(viewModel);

        setStatusBar();

        initView();
    }

    protected abstract void initData();

    protected abstract void initView();

    private void setStatusBar() {
        StatusBarUtil.setColor(this, getResources().getColor(com.bondex.ysl.styleibrary.R.color.colorPrimary));

    }


    /**
     * 创建ViewModel
     *
     * @param cls
     * @param <T>
     * @return
     */
    public <T extends ViewModel> T createViewModel(FragmentActivity activity, Class<T> cls) {
        return ViewModelProviders.of(activity).get(cls);
    }

    @Override
    protected void onStart() {
        super.onStart();

        viewModel.getRefresh().observe(this, refreshObserver);
        viewModel.getMsgLiveDatas().observe(this, msgObserver);

    }

    protected void showLeft(boolean isShow, View.OnClickListener listener) {

        if (ivBack != null && isShow) {
            ivBack.setVisibility(View.VISIBLE);
            ivBack.setOnClickListener(listener);
        }
    }


    protected void showTitle(String title) {

        if (tvTitle != null && !TextUtils.isEmpty(title)) {
            tvTitle.setText(title);
            tvTitle.setVisibility(View.VISIBLE);
        }
    }


    protected void showRight(boolean isShow, int resourceId, View.OnClickListener listener) {

        if (iconRight != null && isShow) {

            iconRight.setVisibility(View.VISIBLE);
            iconRight.setOnClickListener(listener);
            if (resourceId != 0) iconRight.setText(resourceId);
        }
    }

//    处理从viewmodel中传递过来的信息
    protected abstract void handleMessage(Integer msg);

//    listner有关的点击事件
    protected abstract void onMyClick(View view);

    protected void startLoading() {
    }

    protected void stopLoading() {
    }

    protected <T extends ViewModel> T initViewModle() {
        return null;
    }

    protected abstract int getReourceId();


    @Override
    protected void onDestroy() {
        super.onDestroy();

        if (viewModel != null) {


            viewModel.getRefresh().removeObserver(refreshObserver);
            viewModel.getMsgLiveDatas().removeObserver(msgObserver);
            getLifecycle().removeObserver(viewModel);

        }
    }

    protected class MyclickListener extends NoDoubleClickListener {
        @Override
        public void click(View v) {
            onMyClick(v);
        }
    }
}

File: app/src/main/java/com/bondex/ysl/battledore/base/BaseActivity.kt
package com.bondex.ysl.battledore.base

import android.databinding.ViewDataBinding
import android.os.Build
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.support.v7.app.ActionBar
import android.text.TextUtils
import android.view.*
import android.widget.ImageView
import android.widget.RelativeLayout
import android.widget.TextView
import com.bondex.ysl.battledore.R
import com.bondex.ysl.battledore.ui.IconText
import com.bondex.ysl.battledore.util.NoDoubleClickListener
import kotlinx.android.synthetic.main.activity_base.*

abstract class BaseActivity<A:AppCompatActivity,B:ViewDataBinding> : AppCompatActivity() {

    protected val listener = MyClickListener()


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        super.setContentView(R.layout.activity_base)

        setToolBar()

        initData()
    }


    protected fun titleHide(isShow: Boolean){
        base_rl_title?.visibility = if(isShow)View.GONE else View.VISIBLE

    }

    override fun setContentView(layoutResID: Int) {
        super.setContentView(layoutResID)

        val view = layoutInflater.inflate(layoutResID, null)
        val lp: RelativeLayout.LayoutParams = RelativeLayout.LayoutParams(
            RelativeLayout.LayoutParams.MATCH_PARENT,
            RelativeLayout.LayoutParams.MATCH_PARENT
        )
        lp.addRule(RelativeLayout.BELOW, R.id.ll_basetitle_root)

        ll_basetitle_root?.addView(view, lp)
    }

    protected fun setToolBar() {

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
            val flagTranslucentStatus = WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS
            val flagTranslucentNavigation = WindowManager.LayoutParams.FLAG_TRANSLUCENT_NAVIGATION
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                val window = window
                val attributes = window.attributes
                attributes.flags = attributes.flags or flagTranslucentNavigation

                window.attributes = attributes

                getWindow().statusBarColor = resources.getColor(R.color.rect_red)
            } else {
                val window = window
                val attributes = window.attributes
                attributes.flags = attributes.flags or (flagTranslucentStatus or flagTranslucentNavigation)
                window.attributes = attributes
            }
        }

    }



    open fun showLeft(isShow: Boolean, listener: View.OnClickListener?) {

        base_back?.visibility = if (isShow) View.VISIBLE else View.INVISIBLE

        if (listener != null) base_back.setOnClickListener(listener)

    }

    open fun showTitle(isShow: Boolean, titl: String?) {

        base_title?.visibility = if (isShow) View.VISIBLE else View.INVISIBLE

        if (!TextUtils.isEmpty(titl)) base_title?.setText(titl)

    }


    open fun showRight(isShow: Boolean, listener: View.OnClickListener?, resourceId: Int) {

        base_right?.visibility = if (isShow) View.VISIBLE else View.INVISIBLE

        if (listener != null)
            base_right?.setOnClickListener(listener)
        if (resourceId != 0) base_right?.setText(resourceId)
    }

    open abstract fun onCLick(v: View)
    open abstract fun initData()

    protected class MyClickListener : NoDoubleClickListener() {
        override fun click(v: View?) {

            onClick(v)
        }

    }

}
